index.js

```diff
@@ -1,20 +1,69 @@
-export function u(type, props, value) {
-  var node = {type: String(type)}
+/**
+ * @typedef {import('unist').Node} Node
+ * @typedef {import('unist').Parent} Parent
+ * @typedef {import('unist').Literal} Literal
+ * @typedef {Object.<string, unknown>} Props
+ * @typedef {Array.<Node>|string} ChildrenOrValue
+ *
+ * @callback BuildParentWithProps
+ * @param {string} type
+ * @param {Props} props
+ * @param {Array.<Node>} value
+ * @returns {Parent}
+ *
+ * @callback BuildParentWithoutProps
+ * @param {string} type
+ * @param {Array.<Node>} value
+ * @returns {Parent}
+ *
+ * @callback BuildLiteralWithProps
+ * @param {string} type
+ * @param {Props} props
+ * @param {string} value
+ * @returns {Literal}
+ *
+ * @callback BuildLiteralWithoutProps
+ * @param {string} type
+ * @param {string} value
+ * @returns {Literal}
+ *
+ * @callback BuildVoid
+ * @param {string} type
+ * @param {Props} [props]
+ * @returns {Node}
+ */
 
-  if (
-    (value === undefined || value === null) &&
-    (typeof props !== 'object' || Array.isArray(props))
-  ) {
-    value = props
-  } else {
-    Object.assign(node, props)
-  }
+/**
+ * @type {BuildParentWithProps & BuildParentWithoutProps & BuildLiteralWithProps & BuildLiteralWithoutProps & BuildVoid}
+ */
+// prettier-ignore
```

Comment:
Removing this line will make tests pass once, then run prettier and break tests on a second run.

This is why I think CI should check code formatting, not reformat the code and then discard those changes. (`prettier --check .` instead of `prettier --write .`.)

I think using `prettier-ignore` should be used very sparingly (I never used it). This case is legit, because otherwise prettier will break code. Since prettier shouldn’t break code, I think an upstream issue should be created and linked here.

Reply:
How is that different from an `eslint-ignore` for example?

---

this seems close to https://github.com/prettier/prettier/issues/6120 btw. But that fix doesn’t apply on functions.

Reply:
ESLint and Prettier have different purposes.

ESLint checks a variety of rules, of which some may not have a proper resolution. For example a project may require imports to be sorted, but if they use a library which requires one import before another, this can’t be reslved. In this case the rule can be disabled using and `eslint-disable-line` comment.

The goal of Prettier is to provide consistent formatting without breaking any code. This `prettier-ignore` comment is necessary, because not having it makes Prettier break the code. Therefor this comment indicates a bug in Prettier.

I’m not saying this comment shouldn’t be here. I’m saying this is a workaround for a bug in Prettier. I think such workarounds should be paired with a comment linking to an upstream issue.

Reply:
found a way around this: prettier is checking for `@type`, which can also be used

Reply:
Nope...

Reported prettier/prettier#10733.